programs/account-compression/src/instructions/insert_two_leaves_transaction.rs

```diff
@@ -0,0 +1,97 @@
+use std::collections::HashMap;
+
+use anchor_lang::{prelude::*, solana_program::pubkey::Pubkey};
+
+use crate::{
+    emit_indexer_event,
+    state::ConcurrentMerkleTreeAccount,
+    state_merkle_tree_from_bytes_mut,
+    utils::check_registered_or_signer::{GroupAccess, GroupAccounts},
+    ChangelogEvent, ChangelogEventV1, Changelogs, RegisteredProgram,
+};
+
+// TODO: implement group access control
+#[derive(Accounts)]
+pub struct InsertTwoLeavesParallel<'info> {
+    /// CHECK: should only be accessed by a registered program/owner/delegate.
+    #[account(mut)]
+    pub authority: Signer<'info>,
+    pub registered_verifier_pda: Option<Account<'info, RegisteredProgram>>,
+    pub log_wrapper: UncheckedAccount<'info>,
+}
+
+impl GroupAccess for ConcurrentMerkleTreeAccount {
+    fn get_owner(&self) -> &Pubkey {
+        &self.owner
+    }
+
+    fn get_delegate(&self) -> &Pubkey {
+        &self.delegate
+    }
+}
+
+impl<'info> GroupAccounts<'info> for InsertTwoLeavesParallel<'info> {
+    fn get_signing_address(&self) -> &Signer<'info> {
+        &self.authority
+    }
+    fn get_registered_program_pda(&self) -> &Option<Account<'info, RegisteredProgram>> {
+        &self.registered_verifier_pda
+    }
+}
+
+/// for every leaf one Merkle tree account has to be passed as remaing account
+/// for every leaf could be inserted into a different Merkle tree account
+/// 1. deduplicate Merkle trees and identify into which tree to insert what leaf
+/// 2. iterate over every unique Merkle tree and batch insert leaves
+pub fn process_insert_leaves_into_merkle_trees<'a, 'info>(
+    ctx: Context<'a, '_, '_, 'info, InsertTwoLeavesParallel<'info>>,
+    leaves: &'a [[u8; 32]],
+) -> Result<()> {
+    if leaves.len() != ctx.remaining_accounts.len() {
+        return err!(crate::errors::ErrorCode::NumberOfLeavesMismatch);
+    }
+    let mut merkle_tree_map = HashMap::<Pubkey, (&AccountInfo, Vec<&[u8; 32]>)>::new();
+    for (i, mt) in ctx.remaining_accounts.iter().enumerate() {
+        match merkle_tree_map.get(&mt.key()) {
+            Some(_) => {}
+            None => {
+                merkle_tree_map.insert(mt.key(), (mt, Vec::new()));
+            }
+        };
+        merkle_tree_map
+            .get_mut(&mt.key())
+            .unwrap()
```

Comment:
I like accessing members with `.get()` and `.get_mut()` and I think we should do it as often as possible, instead of using the index!

But then, we also shouldn't panic. :stuck_out_tongue: Can we define an error here?

Reply:
agreed, I am using a lot of unwraps all over the pr which we should map to errors!
Not sure this pr is the right one to correct it though.